phase/backup.go

```diff
@@ -0,0 +1,85 @@
+package phase
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/k0sproject/k0sctl/config"
+	"github.com/k0sproject/k0sctl/config/cluster"
+	"github.com/k0sproject/rig/exec"
+	log "github.com/sirupsen/logrus"
+)
+
+var _ phase = &Backup{}
+
+// Backup connect to one of the controllers and takes a backup
+type Backup struct {
+	GenericPhase
+
+	leader *cluster.Host
+}
+
+// Title returns the title for the phase
+func (p *Backup) Title() string {
+	return "Take backup"
+}
+
+// Prepare the phase
+func (p *Backup) Prepare(config *config.Cluster) error {
+	p.Config = config
+	leader := p.Config.Spec.K0sLeader()
+	if leader.Metadata.K0sRunningVersion == "" {
+		return fmt.Errorf("failed to find a running controller")
+	}
+	p.leader = leader
+	return nil
+}
+
+// ShouldRun is true when there is a leader host
+func (p *Backup) ShouldRun() bool {
+	return p.leader != nil
+}
+
+// Run the phase
+func (p *Backup) Run() error {
+	h := p.leader
+	h.Metadata.IsK0sLeader = true
+
+	log.Infof("%s: backing up", h)
+	backupDir, err := h.Configurer.TempDir(h)
+	if err != nil {
+		return err
+	}
+
+	if err := h.Exec(h.K0sBackupCommand(backupDir)); err != nil {
+		return err
+	}
+
+	// get the name of the backup file
+	remoteFile, err := h.ExecOutput(fmt.Sprintf("ls %s", backupDir))
```

Comment:
(btw there's `ExecOutputf`)

Seems like the file is left behind. What happens on second run when `ls` gets two rows? `ExecOutput` will return them both (plus any other file there may have been) as one string and I think it will then try to download a file called "something.tar.gz\nsomething_else.tar.gz" and fail.

Could the filename be somehow parsed from the `k0s backup` output? If not, I guess it should do something like `ls -1t | head -1` to sort by date and only take one.

Reply:
Oh right it goes to a freshly created tempdir.. 🤔